python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+Furthermore for `pyHorace` the second interpreter is not connected with the user's interpreter
+so cannot know about any fitting/modelling functions they have defined.
+
+For these reasons, `pyHorace` opted to use a `mex` [file](https://github.com/mducle/hugo/blob/master/src/call_python.cpp) which is
+called by Matlab instead of the Python function input. 
+This mex file is linked to the CPython API so it can access the parent Python interpreter which is loaded into the same process.
+(Thus this method will not work with an out-of-process Python interpreter).
+
+First when the `DataTypes` class encounters a Python callable (a function or class which can be called), instead of passing this to Matlab,
+it stores the function reference in a [module-global dictionary](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L21),
+and then creates a Matlab [`pythonFunctionWrapper`](https://github.com/mducle/hugo/blob/master/src/pythonFunctionWrapper.m) object
+which stores the (string) key to the Python function.
+Whenever `call_method` encounters a `pythonFunctionWrapper` object,
+it [wraps](https://github.com/mducle/hugo/blob/master/src/check_wrapped_function.m) this in a call to the mex `call_python` function with the correct key.
+Thus it appears to the user that Matlab is calling a Python function transparently (but actually happens is that Matlab calls `call_python`).
+`call_python` then [retrieves](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L390) the Python function reference from the module-global `dict` 
+and [calls](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L413) that function with the arguments given to it.
+
+The majority of the code in `call_python` is for data conversion between Matlab and Python.
+This is because `mex` functions may only take and return *Matlab* data types (Python objects are not allowed).
+Similarly the CPython `PyObject_Call` function requires that inputs to the called functions are all of the `PyObject` type.
+Thus `call_python` converts Matlab types to their Python equivalent and vice versa as per this table:
+
+|      Matlab    |      Python         | Note |
+| -------------- | ------------------- | ---- |
+| numeric array  | `numpy.ndarray`     |      |
+| numeric array  | `list` of numbers   | p2m  |
+| `char` array   | `str`               |      |
+| `string`       | `str`               |      |
+| `string` array | `list` of `str`     |      |
+| `struct`       | `dict`              |      |
+| `struct` array | `list` of `dict`    |      |
+| `cell` array   | `list`              |      |
+| `cell` array   | `tuple`             | p2m  |
+
+"p2m" indicates that conversion only applies from Python to Matlab 
+(from a Python `list` of numbers to a Matlab numeric array, and from a python `tuple` to a Matlab `cell` array).
+Other conversions apply in both directions (e.g. Matlab cell arrays are always converted to a Python `list` and vice versa).
```

Comment:
Is this consistent? cell_array always converted to list and v.v., but with p2m cell_array is converted to tuple.  One of these must be wrong.

Reply:
So going from Matlab to Python I chose to always convert a cell array to a `list`.

Going from Python to Matlab both `tuple` and `list` are converted to cell arrays.

E.g. this converter will never *create* a Python `tuple` from any Matlab types but accepts `tuple`s as inputs. 

This is because Matlab doesn't have any equivalent to a tuple.